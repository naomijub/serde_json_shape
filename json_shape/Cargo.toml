
[package]
name = "json_shape"
version = "0.4.1"
edition = "2024"
authors = ["Julia Naomi <jnboeira@outlook.com>",]
description = "Rust library to handle JSON shape"
readme = "README.md"
license = "Apache-2.0"
homepage = "https://github.com/naomijub/serde_json_shape"
repository = "https://github.com/naomijub/serde_json_shape/json_shape"
keywords = ["JSON", "serde", "shape", "serialization"]
categories = ["encoding", "parser-implementations"]


[dependencies]
serde = { version = "1.0.194", features = ["derive"] }
serde_json = "1.0"
thiserror = "2.0"
logos = "0.15"
codespan-reporting = "0.12"

[dev-dependencies]
insta = { version = "1.43.1", features = [] }

[build-dependencies]
lelwel = "0.8.0"

[lints]
workspace = true