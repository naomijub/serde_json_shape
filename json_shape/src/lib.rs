#![doc = include_str!("../../README.md")]
#![allow(clippy::redundant_pub_crate)]
/// Module containing Error types
pub mod error;
mod value;

mod lexer;
pub(crate) mod parser;
/// [`serde_json`] related functions and types
pub mod serde;
pub(crate) mod shape;

use std::str::FromStr;

use crate::{
    error::Error,
    parser::Parser,
    shape::{merger::merge, parse_cst},
    value::Value,
};

pub use value::Similar;
pub use value::Value as JsonShape;

/// Creates a [`JsonShape`] from a single Json source
/// ```
/// use json_shape::JsonShape;
/// use std::str::FromStr;
///
/// let source = "[12, 34, 56]";
/// let json_shape = JsonShape::from_str(source).unwrap();
/// ```
impl FromStr for Value {
    type Err = Error;

    fn from_str(source: &str) -> Result<Self, Self::Err> {
        let cst = Parser::parse(source, &mut Vec::new());

        let value = parse_cst(&cst, source)?;

        Ok(value)
    }
}

impl Value {
    /// Creates a [`JsonShape`] from multiple Json sources
    ///
    /// # Errors
    ///
    /// Will return `Err` if failed to parse Json or if shapes don't align.
    pub fn from_sources(sources: &[&str]) -> Result<Self, Error> {
        let mut diags = Vec::new();
        let mut values = Vec::new();
        for source in sources {
            let cst = Parser::parse(source, &mut diags);

            values.push(parse_cst(&cst, source)?);
        }

        merge(&values)
    }

    /// Checks if Json is subset of specific [`JsonShape`]
    /// ```rust
    /// use std::str::FromStr;
    ///
    /// use json_shape::{IsSubset, JsonShape};
    /// let shape = JsonShape::Object { content: [
    ///     ("name".to_string(), JsonShape::String { optional: false }),
    ///     ("surname".to_string(), JsonShape::String { optional: false }),
    ///     ("middle name".to_string(), JsonShape::String { optional: true }),
    ///     ("age".to_string(), JsonShape::Number { optional: false }),
    ///     ("id".to_string(), JsonShape::OneOf { variants: [
    ///         JsonShape::Object { content: [
    ///             ("number".to_string(), JsonShape::Number { optional: false }),
    ///             ("state".to_string(), JsonShape::String { optional: false }),
    ///         ].into(), optional: false },
    ///         JsonShape::Array { r#type: Box::new(JsonShape::Number { optional: false }), optional: false }
    ///     ].into(), optional: false })
    /// ].into(), optional: false };
    ///
    /// let json = r#"{
    /// "name": "lorem",
    /// "surname": "ipsum",
    /// "age": 30,
    /// "id": {
    ///     "number": 123456,
    ///     "state": "st"
    /// }
    /// }"#;
    ///
    /// let shape_1 = JsonShape::from_str(json).unwrap();
    ///
    /// # assert!(
    /// shape_1.is_subset(&shape)
    /// # );
    /// # assert!(
    /// shape.is_superset(json)
    /// # );
    /// ```
    #[must_use]
    pub fn is_superset(&self, json: &str) -> bool {
        let Ok(value) = Self::from_str(json) else {
            return false;
        };

        value.is_subset(self)
    }

    /// Checks if Json is subset of specific [`JsonShape`]
    ///
    /// - Checked version of [`is_superset`]
    ///
    /// # Errors
    ///
    /// Returns `Err` if failed to parse Json
    pub fn is_superset_checked(&self, json: &str) -> Result<bool, Error> {
        let value = Self::from_str(json)?;

        Ok(value.is_subset(self))
    }
}

/// Determines if `T::self` is subset of `T`.
/// Only implemented for [`JsonShape`]/[`Value`]. A few rule examples
///
/// - `JsonShape::Number` is subset of `JsonShape::Option<Number>`
/// - `JsonShape::Null` is subset of `JsonShape::Option<Number>` and  `JsonShape::Null`
/// - `JsonShape::Number` is subset of `JsonShape::OneOf[Number | String]`
/// - `JsonShape::Number` is *NOT* subset of `JsonShape::Array<Number>` => `1.23 != [1.23]`
/// - `JsonShape::Array<Number>` is subset of `JsonShape::Array<OnOf<[Number | Boolean]>>`
/// - `JsonShape::Object{"key_a": JsonShape::Number}` is *NOT* subset of `JsonShape::Object{"key_b": JsonShape::Number}` => `key_a != key_b`
/// - `JsonShape::Object{"key_a": JsonShape::Number}` is subset of `JsonShape::Object{"key_a": JsonShape::Option<Number>}`
/// - `JsonShape::Object{"key_a": JsonShape::Number}` is subset of `JsonShape::Object{"key_a": JsonShape::OneOf[Number | Boolean]}`
pub trait IsSubset {
    /// Determines if `T::self` is subset of other `T`
    fn is_subset(&self, other: &Self) -> bool;
}
