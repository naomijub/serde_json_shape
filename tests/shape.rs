#![allow(missing_docs)]

use std::str::FromStr;

use insta::assert_snapshot;
use json_shape::{JsonShape, error::Error};

#[test]
fn parse_nil_shape() {
    let source = "null";
    let value = JsonShape::from_str(source).unwrap();

    assert_eq!(value, JsonShape::Null);
}

#[test]
fn parse_true_shape() {
    let source = "true";
    let value = JsonShape::from_str(source).unwrap();

    assert_eq!(value, JsonShape::Bool { optional: false });
}

#[test]
fn parse_false_shape() {
    let source = "false";
    let value = JsonShape::from_str(source).unwrap();

    assert_eq!(value, JsonShape::Bool { optional: false });
}

#[test]
fn parse_number_shape() {
    let source = "123.456";
    let value = JsonShape::from_str(source).unwrap();

    assert_eq!(value, JsonShape::Number { optional: false });
}

#[test]
fn parse_string_shape() {
    let source = "\"this is a string\"";
    let value = JsonShape::from_str(source).unwrap();

    assert_eq!(value, JsonShape::String { optional: false });
}

#[test]
fn parse_something_else_shape_error() {
    let source = "random";
    let err = JsonShape::from_str(source).unwrap_err();

    assert_eq!(err, Error::InvalidJson("random".to_string(), 0..6));
}

#[test]
fn parse_array_shape() {
    let source = "[\"string\", 123.456, 234, true, false, null]";
    let value = JsonShape::from_str(source).unwrap();

    assert_eq!(
        value,
        JsonShape::Tuple {
            elements: vec![
                JsonShape::String { optional: false },
                JsonShape::Number { optional: false },
                JsonShape::Number { optional: false },
                JsonShape::Bool { optional: false },
                JsonShape::Bool { optional: false },
                JsonShape::Null,
            ],
            optional: false
        }
    );
}

#[test]
fn parse_map_shape() {
    let source = "{
    \"a\" : 1234,
    \"b\": true
    }";
    let value = JsonShape::from_str(source).unwrap();
    let content = [
        (
            String::from_str("a").unwrap(),
            JsonShape::Number { optional: false },
        ),
        (
            String::from_str("b").unwrap(),
            JsonShape::Bool { optional: false },
        ),
    ]
    .into();
    assert_eq!(
        value,
        JsonShape::Object {
            content,
            optional: false
        }
    );
}

#[test]
fn complex_json_shape() {
    let source = r#"{
        "str": "this is a string",
        "number": 123.456,
        "bool_true": true,
        "bool_false": false,
        "nil": null,
        "array": [1, 2, 3, 4],
        "tuple": [123, "string", true],
        "map": {
          "a": "b",
          "c": 123
        },
        "array of maps": [
            {
                "a": "b",
                "c": 123
            },
            {
                "a": "b",
                "b": true
            }
        ]
    }"#;

    let value = JsonShape::from_str(source).unwrap();
    assert_eq!(
        value,
        JsonShape::Object {
            content: [
                (
                    "array".to_string(),
                    JsonShape::Array {
                        r#type: Box::new(JsonShape::Number { optional: false }),
                        optional: false
                    }
                ),
                (
                    "tuple".to_string(),
                    JsonShape::Tuple {
                        elements: vec![
                            JsonShape::Number { optional: false },
                            JsonShape::String { optional: false },
                            JsonShape::Bool { optional: false }
                        ],
                        optional: false
                    },
                ),
                (
                    "array of maps".to_string(),
                    JsonShape::Array {
                        r#type: Box::new(JsonShape::Object {
                            content: [
                                ("a".to_string(), JsonShape::String { optional: false }),
                                ("c".to_string(), JsonShape::Number { optional: true }),
                                ("b".to_string(), JsonShape::Bool { optional: true })
                            ]
                            .into(),
                            optional: false
                        }),
                        optional: false
                    }
                ),
                (
                    "bool_false".to_string(),
                    JsonShape::Bool { optional: false },
                ),
                ("bool_true".to_string(), JsonShape::Bool { optional: false },),
                (
                    "map".to_string(),
                    JsonShape::Object {
                        content: [
                            ("a".to_string(), JsonShape::String { optional: false }),
                            ("c".to_string(), JsonShape::Number { optional: false })
                        ]
                        .into(),
                        optional: false
                    }
                ),
                ("nil".to_string(), JsonShape::Null,),
                ("number".to_string(), JsonShape::Number { optional: false },),
                ("str".to_string(), JsonShape::String { optional: false })
            ]
            .into(),
            optional: false
        }
    );
    assert_snapshot!(value);
}

#[test]
fn parse_duplicated_keys_object_shape() {
    let source = "{\"a\": 123, \"a\": true}";
    let err = JsonShape::from_str(source).unwrap_err();

    assert_eq!(
        err.to_string(),
        "invalid type `Boolean`. Expected `Number`."
    );
}

#[test]
fn complex_json_shape_from_sources() {
    let source_1 = r#"{
        "str": "this is a string",
        "number": 123.456,
        "nil": null,
        "array": [123, "string", true],
        "map": {
          "a": "b",
          "c": 123,
          "d": 1
        },
        "array of maps": [
            {
                "a": "b",
                "c": 123
            }
        ]
    }"#;

    let source_2 = r#"{
        "str": "this is a string",
        "bool": true,
        "nil": null,
        "array": [123, "string"],
        "map": {
          "a": "b",
          "c": 123,
          "e": 2
        },
        "array of maps": [
            {
                "a": "b",
                "b": true
            }
        ]
    }"#;

    let shape = JsonShape::from_sources(&[source_1, source_2]).unwrap();

    assert_snapshot!(shape);
}
