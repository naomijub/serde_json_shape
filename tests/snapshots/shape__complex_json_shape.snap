---
source: tests/shape.rs
expression: value
---
Object{"array": Array<OneOf[Boolean | Number | String]>, "array of maps": Array<Object{"a": String, "b": Option<Boolean>, "c": Option<Number>}>, "bool_false": Boolean, "bool_true": Boolean, "map": Object{"a": String, "c": Number}, "nil": Null, "number": Number, "str": String}
